### ISSUE TITLE
CRF layer negative loss

### ISSUE THREAD

**Issue by mdetournay** [keras-team/keras-contrib#253]
Hi, 

I am training an LSTM - CRF network for named entity recognition. When using crf.loss_function, I'm getting negative losses after a few epochs. Before I go on giving more details about my code, is this even possible with this crf.loss_function? I tried looking at  the code but with this recursive code I'm having trouble identifying where I could be getting negative values. 

Thank you very much for your help! 

Cheers

Martin 

Here is the network: 

`input=Input(shape=(170,))
        x=Embedding(input_dim=len(TrainingObj.keras_tokenizer.word_index)+1,output_dim=100,weights=[TrainingObj.embedding_matrix],
                    mask_zero=False,input_length=170,trainable=False)(input)
        x=Bidirectional(GRU(units=64,return_sequences=True,dropout=0.1),merge_mode='concat')(x)
        x=TimeDistributed(Dense(64, activation='relu'))(x)

        crf=CRF(num_classes,sparse_target=False)
        out=crf(x)

        self.model=Model(inputs=input,outputs=out)
        if restart:
            self.load_model()

        self.model.compile(optimizer='adam', loss=crf.loss_function, metrics=[crf.accuracy])
`
Layer (type)                 Output Shape              Param #   
=================================================================
input_1 (InputLayer)         (None, 170)               0         
_________________________________________________________________
embedding_1 (Embedding)      (None, 170, 100)          22203800  
_________________________________________________________________
bidirectional_1 (Bidirection (None, 170, 128)          63360     
_________________________________________________________________
time_distributed_1 (TimeDist (None, 170, 64)           8256      
_________________________________________________________________
crf_1 (CRF)                  (None, 170, 3)            210       
=================================================================
Total params: 22,275,626
Trainable params: 71,826
Non-trainable params: 22,203,800

**Comment by alexpython1988** [comment]
I have the same problem. I tried to mask the padding 0s, but no luck. I wonder if this issue is related to the position where we pad the 0. I padded the 0 in front of each training sample, I wonder if anyone tried post padding?

**Comment by Euruson** [comment]
Is your backend Theano? Maybe you can see #131.

**Comment by sharonrapoport** [comment]
I have the same issue, did you end up figuring this out?

**Comment by lzfelix** [comment]
Not yet, as I’m masking my inputs... But it should be fixed..

**Comment by lzfelix** [comment]
By the way, even if you don’t need masking, I have found that adding a
dummy mask layer after the input layer fixes this issue...

**Comment by EvilPort2** [comment]
> By the way, even if you don’t need masking, I have found that adding a
dummy mask layer after the input layer fixes this issue...

@lzfelix How do I do that? I am new to this thing. Currently this is my code.
```python
def lstm_model():
	input_text = Input(shape=(None, ), dtype="string", name='input_layer')
	embedding = Lambda(get_elmo, name='elmo_embedding')(input_text)
	x = Bidirectional(LSTM(units=100, return_sequences=True, recurrent_dropout=0.5, dropout=0.5), name='bilstm1')(embedding)
	x = Bidirectional(LSTM(units=100, return_sequences=True, recurrent_dropout=0.5, dropout=0.5), name='bilstm2')(x)
	x = TimeDistributed(Dense(50, activation="relu"))(x)
	crf = CRF(no_of_tags, sparse_target=True)
	out = crf(x)
	model = Model(input_text, out)
	return model, crf
model, crf = lstm_model()
model.compile(optimizer='rmsprop', loss=crf.loss_function, metrics=[crf.accuracy])
model.fit(np.array(train_x), train_y, validation_data=(np.array(valid_x), valid_y), shuffle=True, batch_size=batch_size, epochs=1, verbose=1, callbacks=[checkpoint])
```

**Comment by JuntingGuo** [comment]
I encountered this problem today and I realized that its because I used wrong config for CRF layer. My y_train is in one-hot form but I specified `sparse_target=True` in the CRF layer therefore the the layer chose the wrong loss function and gave negative loss. Hope it helps.

**Comment by yongzhuo** [comment]
I have the same issue, when I am training an BERT + Bi-LSTM + CRF network for named entity recognition of chinese_people_daily. how could I slove it?
some logs as follows


20800/20864 [============================>.] - ETA: 0s - loss: -0.6666 - crf_accuracy: 0.9548
20816/20864 [============================>.] - ETA: 0s - loss: -0.6667 - crf_accuracy: 0.9548
20832/20864 [============================>.] - ETA: 0s - loss: -0.6667 - crf_accuracy: 0.9548
20848/20864 [============================>.] - ETA: 0s - loss: -0.6667 - crf_accuracy: 0.9548
20864/20864 [==============================] - 107s 5ms/step - loss: -0.6667 - crf_accuracy: 0.9548 - val_loss: -0.6612 - val_crf_accuracy: 0.9498

**Comment by tapos12** [comment]
I am experiencing same issue for CRF layer

**Comment by tapos12** [comment]
> > By the way, even if you don’t need masking, I have found that adding a
> > dummy mask layer after the input layer fixes this issue...
> 
> @lzfelix How do I do that? I am new to this thing. Currently this is my code.
> 
> ```python
> def lstm_model():
> 	input_text = Input(shape=(None, ), dtype="string", name='input_layer')
> 	embedding = Lambda(get_elmo, name='elmo_embedding')(input_text)
> 	x = Bidirectional(LSTM(units=100, return_sequences=True, recurrent_dropout=0.5, dropout=0.5), name='bilstm1')(embedding)
> 	x = Bidirectional(LSTM(units=100, return_sequences=True, recurrent_dropout=0.5, dropout=0.5), name='bilstm2')(x)
> 	x = TimeDistributed(Dense(50, activation="relu"))(x)
> 	crf = CRF(no_of_tags, sparse_target=True)
> 	out = crf(x)
> 	model = Model(input_text, out)
> 	return model, crf
> model, crf = lstm_model()
> model.compile(optimizer='rmsprop', loss=crf.loss_function, metrics=[crf.accuracy])
> model.fit(np.array(train_x), train_y, validation_data=(np.array(valid_x), valid_y), shuffle=True, batch_size=batch_size, epochs=1, verbose=1, callbacks=[checkpoint])
> ```

Did you figure it out while using ELMo?

**Comment by moemmur** [comment]
Any solutions for this? The research says that crf at the end for the sequence labelling task performs better than Softmax layer, however, in the case of Named Entity Recognition I have seen worst results with crf. And I also experience the - ve loss. Any solution for this? Did anybody found the solution?

**Comment by deanhoperobertson** [comment]
> > By the way, even if you don’t need masking, I have found that adding a
> > dummy mask layer after the input layer fixes this issue...
> 
> @lzfelix How do I do that? I am new to this thing. Currently this is my code.
> 
> ```python
> def lstm_model():
> 	input_text = Input(shape=(None, ), dtype="string", name='input_layer')
> 	embedding = Lambda(get_elmo, name='elmo_embedding')(input_text)
> 	x = Bidirectional(LSTM(units=100, return_sequences=True, recurrent_dropout=0.5, dropout=0.5), name='bilstm1')(embedding)
> 	x = Bidirectional(LSTM(units=100, return_sequences=True, recurrent_dropout=0.5, dropout=0.5), name='bilstm2')(x)
> 	x = TimeDistributed(Dense(50, activation="relu"))(x)
> 	crf = CRF(no_of_tags, sparse_target=True)
> 	out = crf(x)
> 	model = Model(input_text, out)
> 	return model, crf
> model, crf = lstm_model()
> model.compile(optimizer='rmsprop', loss=crf.loss_function, metrics=[crf.accuracy])
> model.fit(np.array(train_x), train_y, validation_data=(np.array(valid_x), valid_y), shuffle=True, batch_size=batch_size, epochs=1, verbose=1, callbacks=[checkpoint])
> ```
@EvilPort2 you should only use  sparse_target=True if your labels are NOT one-hot encoded.

**Comment by EvilPort2** [comment]
> > > By the way, even if you don’t need masking, I have found that adding a
> > > dummy mask layer after the input layer fixes this issue...
> > 
> > 
> > @lzfelix How do I do that? I am new to this thing. Currently this is my code.
> > ```python
> > def lstm_model():
> > 	input_text = Input(shape=(None, ), dtype="string", name='input_layer')
> > 	embedding = Lambda(get_elmo, name='elmo_embedding')(input_text)
> > 	x = Bidirectional(LSTM(units=100, return_sequences=True, recurrent_dropout=0.5, dropout=0.5), name='bilstm1')(embedding)
> > 	x = Bidirectional(LSTM(units=100, return_sequences=True, recurrent_dropout=0.5, dropout=0.5), name='bilstm2')(x)
> > 	x = TimeDistributed(Dense(50, activation="relu"))(x)
> > 	crf = CRF(no_of_tags, sparse_target=True)
> > 	out = crf(x)
> > 	model = Model(input_text, out)
> > 	return model, crf
> > model, crf = lstm_model()
> > model.compile(optimizer='rmsprop', loss=crf.loss_function, metrics=[crf.accuracy])
> > model.fit(np.array(train_x), train_y, validation_data=(np.array(valid_x), valid_y), shuffle=True, batch_size=batch_size, epochs=1, verbose=1, callbacks=[checkpoint])
> > ```
> 
> @EvilPort2 you should only use sparse_target=True if your labels are NOT one-hot encoded.

I had figured it out way earlier. But thanks for your reply anyway.


> > > By the way, even if you don’t need masking, I have found that adding a
> > > dummy mask layer after the input layer fixes this issue...
> > 
> > 
> > @lzfelix How do I do that? I am new to this thing. Currently this is my code.
> > ```python
> > def lstm_model():
> > 	input_text = Input(shape=(None, ), dtype="string", name='input_layer')
> > 	embedding = Lambda(get_elmo, name='elmo_embedding')(input_text)
> > 	x = Bidirectional(LSTM(units=100, return_sequences=True, recurrent_dropout=0.5, dropout=0.5), name='bilstm1')(embedding)
> > 	x = Bidirectional(LSTM(units=100, return_sequences=True, recurrent_dropout=0.5, dropout=0.5), name='bilstm2')(x)
> > 	x = TimeDistributed(Dense(50, activation="relu"))(x)
> > 	crf = CRF(no_of_tags, sparse_target=True)
> > 	out = crf(x)
> > 	model = Model(input_text, out)
> > 	return model, crf
> > model, crf = lstm_model()
> > model.compile(optimizer='rmsprop', loss=crf.loss_function, metrics=[crf.accuracy])
> > model.fit(np.array(train_x), train_y, validation_data=(np.array(valid_x), valid_y), shuffle=True, batch_size=batch_size, epochs=1, verbose=1, callbacks=[checkpoint])
> > ```
> 
> Did you figure it out while using ELMo?

Yes. I figured out that while I was using ELMO

**Comment by Tangqian1994** [comment]
> I have the same issue, when I am training an BERT + Bi-LSTM + CRF network for named entity recognition of chinese_people_daily. how could I slove it?
> some logs as follows
> 
> 20800/20864 [============================>.] - ETA: 0s - loss: -0.6666 - crf_accuracy: 0.9548
> 20816/20864 [============================>.] - ETA: 0s - loss: -0.6667 - crf_accuracy: 0.9548
> 20832/20864 [============================>.] - ETA: 0s - loss: -0.6667 - crf_accuracy: 0.9548
> 20848/20864 [============================>.] - ETA: 0s - loss: -0.6667 - crf_accuracy: 0.9548
> 20864/20864 [==============================] - 107s 5ms/step - loss: -0.6667 - crf_accuracy: 0.9548 - val_loss: -0.6612 - val_crf_accuracy: 0.9498

hello, have you figure it out? the negetive loss issue

**Comment by NirmalenduPrakash** [comment]
I have the same issue. The loss is limited to small negative value. This loss should never go negative(https://arxiv.org/pdf/1603.01360.pdf). I think this is because of logsumexp(doind direct exp and sum can be unstable) approximation.This anyways this doesn't matter because following the gradient can only lead to best path probability(emission+transition) to be equal to sum of all probabilities(all paths), only slightly negative in this case due to approximation.

**Comment by waring92** [comment]
I think because of activation function in crf layer used in calculating negative log likelihood scalar.
Add 'Relu' activation function of keras where loss is calculated.
then, i solved this problem without any changes in prediction performance

```
    def loss(y_true, y_pred):
        X = crf.input
        mask = crf.input_mask
        nloglik = crf.get_negative_log_likelihood(y_true, X, mask)
        return tf.keras.activations.relu(nloglik)
```

**Comment by NirmalenduPrakash** [comment]
There was a bug in my code. It doesn't give negative loss now.

**Comment by pinesnow72** [comment]
I've found a solution of this issue. Refer to [my comment](https://github.com/keras-team/keras-contrib/issues/278#issuecomment-680381935)

**Comment by cpmss521** [comment]
I have the same issue，  how could I slove it?

**Comment by pinesnow72** [comment]
> I have the same issue， how could I slove it?

Please refer to [this link](https://github.com/keras-team/keras-contrib/issues/278#issuecomment-680381935)

**Comment by cpmss521** [comment]
i confront the same problem, anybody has a  solution?
